### PR TITLE
Use thiserror for credential provider errors

### PR DESCRIPTION
**Pull request by arlosi**
### What does this PR try to resolve?
Errors from credential providers currently must a single string. This leads to a lot of `.map_err(|e|cargo_credential::Error::Other(e.to_string())`, which loses the `source()` of these errors.

This changes the `cargo_credential::Error` to use `thiserror` and adds a custom serialization for `std::error::Error` that preserves the source error chain across serialization / deserialization.

A unit test is added to verify serialization / deserialization.

### CODE REVIEW

File: credential/cargo-credential/Cargo.toml [hunk 1]
```diff
@@ -7,6 +7,8 @@ repository = "https://github.com/rust-lang/cargo"
 description = "A library to assist writing Cargo credential helpers."
 
 [dependencies]
+anyhow.workspace = true
```
Comment (arlosi) [inline credential/cargo-credential/Cargo.toml:10]: `anyhow` could be made an optional dependency with a feature to enable it. It's here to enable a convenience conversion.

File: credential/cargo-credential/src/error.rs [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+/// 
+/// `UrlNotSupported` and `NotFound` errors both cause Cargo
+/// to attempt another provider, if one is available. The other 
+/// variants are fatal.
+/// 
+/// Note: Do not add a tuple variant, as it cannot be serialized.
+#[derive(Serialize, Deserialize, ThisError, Debug)]
+#[serde(rename_all = "kebab-case", tag = "kind")]
+#[non_exhaustive]
+pub enum Error {
```
Comment (epage) [inline credential/cargo-credential/src/error.rs:15]: For Deserialization, do we need to worry about older credential providers having more `Error` variants?

Comment (arlosi) [reply credential/cargo-credential/src/error.rs:15]: I think we're OK making a breaking change by removing variants here since it's unstable and it was only added last week.

File: credential/cargo-credential/src/error.rs [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+    UrlNotSupported,
+    #[error("credential not found")]
+    NotFound,
+    #[error("requested operation not supported")]
+    OperationNotSupported,
+    #[error("protocol version {version} not supported")]
+    ProtocolNotSupported { version: u32 },
+    #[error(transparent)]
+    #[serde(with = "error_serialize")]
+    Other(Box<dyn StdError + Sync + Send>),
```
[the remark this answers is not in the document]
Comment (epage) [reply credential/cargo-credential/src/error.rs:26]: And whether messages and programmatic variants are actually mutually exclusive or not

[the remark this answers is not in the document]
Comment (epage) [reply credential/cargo-credential/src/error.rs:26]: Oh, most of this moved so I guess this is out of scope for this PR though something for us to keep in mind

File: credential/cargo-credential/src/lib.rs [hunk 4]
```diff
@@ -63,7 +65,7 @@ pub struct RegistryInfo<'a> {
     /// The crates.io registry will be `crates-io` (`CRATES_IO_REGISTRY`).
     pub name: Option<&'a str>,
     /// Headers from attempting to access a registry that resulted in a HTTP 401.
-    #[serde(skip_serializing_if = "Vec::is_empty")]
+    #[serde(skip_serializing_if = "Vec::is_empty", default)]
```
Comment (epage) [inline credential/cargo-credential/src/lib.rs:68]: Can we at least split this unrelated fix into a separate commit?

File: credential/cargo-credential/Cargo.toml (hunk 1 above)
Comment (epage) [reply credential/cargo-credential/Cargo.toml:10]: Not the biggest fan of things like `anyhow` being in library APIs **but** the place we care most about compatibility is the wire protocol and `anyhow` doesn't matter there so ... meh.

File: credential/cargo-credential/src/error.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (arlosi) [reply credential/cargo-credential/src/error.rs:26]: I wanted to have variants where Cargo may want to do something differently based on the provider output. Everything else should go into the `Other` variant.

`UrlNotSupported` and `NotFound` both cause Cargo to try the next provider in the global credential provider list rather than erroring out.

I did reduce the number of variants in this change, but we could possibly reduce it further. I kept `ProtocolNotSupported` and `OperationNotSupported` in case we wanted to handle those differently in the future.

File: credential/cargo-credential/src/error.rs [hunk 5]
```diff
@@ -0,0 +1,193 @@
+use serde::{Deserialize, Serialize};
+use std::error::Error as StdError;
+use thiserror::Error as ThisError;
+
+/// Credential provider error type.
+/// 
+/// `UrlNotSupported` and `NotFound` errors both cause Cargo
+/// to attempt another provider, if one is available. The other 
+/// variants are fatal.
```
Comment (epage) [inline credential/cargo-credential/src/error.rs:9]: By making it easier to use, a part of me worries that we are making it harder to correctly report the right error variant.

The way I've handled this in the error codes for my CLI is to have a sloppy error type for the implementation of each function call in `main` and then have a more constrained error type with explicit conversions used by `main`.  This does require the structure of the code to align well with the error variants.

**Review by epage — COMMENTED** [review]
Thanks!  Overall, I appreciate how much this cleans up the code

File: credential/cargo-credential/src/error.rs (hunk 2 above)
Comment (epage) [reply credential/cargo-credential/src/error.rs:15]: I meant for when this becomes stable, what constraints are we putting on evolution and how do we reduce them.

File: credential/cargo-credential/src/error.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (epage) [reply credential/cargo-credential/src/error.rs:26]: Makes sense.  The fact that we only document behavior differences with some variants is partly why I asked.

File: credential/cargo-credential/src/error.rs (hunk 5 above)
Comment (arlosi) [reply credential/cargo-credential/src/error.rs:9]: Thanks. I'm definitely open to feedback on improving this. 

Since it will be implemented by external crates, it's hard to know up front what all the variants should be. In the original implementation, I tried to anticipate the common ones and add variants, but in this change I decided to reduce the number of variants and have an `Other` case.

I'm not quite sure what you're suggesting here. Removing some of the `From` impls that convert to `Other` and requiring explicit conversions?

Comment (epage) [reply credential/cargo-credential/src/error.rs]: > I'm not quite sure what you're suggesting here. Removing some of the From impls that convert to Other and requiring explicit conversions?

Not quite a suggestion but effectively yes.  Phrased differently: "I'm looking for input on whether we should reduce `From` impls to encourage explicit conversions so people are more likely to specify the correct error semantics".

Comment (arlosi) [reply credential/cargo-credential/src/error.rs]: I made changes to remove the `From` impls for `serde::Error` and `io::Error` and added more error contexts.

File: credential/cargo-credential/src/error.rs (hunk 2 above)
Comment (arlosi) [reply credential/cargo-credential/src/error.rs:15]: Great point! I added a new `Unknown` variant marked with `serde(other)` to each of the enums that's serialized across the JSON protocol. That enables deserialization to succeed even if there are new variants added. Between this and marking the enums `non_exhaustive`, we're hopefully flexible enough.

File: credential/cargo-credential/src/error.rs [hunk 6]
```diff
@@ -13,29 +13,32 @@ use thiserror::Error as ThisError;
...
 
-impl From<serde_json::Error> for Error {
-    fn from(err: serde_json::Error) -> Self {
-        Box::new(err).into()
-    }
+    /// A new variant was added to this enum since Cargo was built
+    #[error("unknown error kind; try updating Cargo?")]
+    #[serde(other)]
+    Unknown,
 }
```
Comment (epage) [inline credential/cargo-credential/src/error.rs]: nit: was this meant to be part of the next commit?

File: credential/cargo-credential/src/error.rs [hunk 7]
```diff
@@ -156,6 +159,16 @@ mod error_serialize {
...
 
+    #[test]
+    pub fn unknown_kind() {
+        let json = r#"{
+            "kind": "unexpected-kind",
+            "unexpected-content": "test"
+          }"#;
+        let e: Error = serde_json::from_str(&json).unwrap();
+        assert!(matches!(e, Error::Unknown));
+    }
```
Comment (epage) [inline credential/cargo-credential/src/error.rs:170]: nit: was this meant to be part of the next commit?

File: credential/cargo-credential/src/error.rs (hunk 5 above)
Comment (epage) [reply credential/cargo-credential/src/error.rs]: To be clear, I wasn't saying it had to be done but was wanting to make sure we made a conscious decision.  If this is what you feel is best great.  If not, feel free to say so.

**Comment by epage** [comment]
Overall, I'm fine with it as-is and leave it to arlosi if any of the remaining points are changed.

@bors r=arlosi

**Comment by epage** [comment]
@bors cancel

I got the command wrong.  Feel free to r= me
